### PR TITLE
Continue refactoring macro expansion and resolution

### LINKED ISSUE #44692: Hangs on derive macro invoked by format string macro
The compiler hangs when compiling the following code.

```rust
macro_rules! hang {
    () => {
        {
            #[derive(Clone)]
            struct S;

            ""
        }
    }
}

fn main() {
    format_args!(hang!());
}
```

This is a regression in 1.20.0. The same code terminates with an error "format argument must be a string literal" in 1.19.0, which is the correct behavior.

### LINKED ISSUE #52363: Expansion info is reset for the mark
```rust
format_args!({ #[derive(repro)] struct ProcMacroHack; });
```

Correct behavior on all compilers 1.15.0 through 1.28.0-beta.10:

```
error: format argument must be a string literal.
 --> src/main.rs:5:18
  |
5 |     format_args!({ #[derive(repro)] struct ProcMacroHack; });
  |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

But on nightly:
(UPDATE: The ICE is now suppressed, but the issue still needs investigation.)
```
thread 'main' panicked at 'expansion info is reset for the mark 2
old: ExpnInfo {
    call_site: Span {
        lo: BytePos(
            56
        ),
        hi: BytePos(
            113
        ),
        ctxt: #0
    },
    def_site: None,
    format: MacroBang(
        format_args
    ),
    allow_internal_unstable: true,
    allow_internal_unsafe: false,
    local_inner_macros: false,
    edition: Edition2015
}
new: ExpnInfo {
    call_site: Span {
        lo: BytePos(
            88
        ),
        hi: BytePos(
            109
        ),
        ctxt: #0
    },
    def_site: None,
    format: MacroAttribute(
        derive(repro)
    ),
    allow_internal_unstable: true,
    allow_internal_unsafe: false,
    local_inner_macros: false,
    edition: Edition2015
}', libsyntax_pos/hygiene.rs:114:17
note: Run with `RUST_BACKTRACE=1` for a backtrace.

error: internal compiler error: unexpected panic
```

---

#### Repro script

```sh
#!/bin/sh

cargo new repro
cargo new --lib repro_macro

cat >repro/src/main.rs <<'REPRO'
#[macro_use]
extern crate repro_macro;

fn main() {
    format_args!({ #[derive(repro)] struct ProcMacroHack; });
}
REPRO

cat >>repro/Cargo.toml <<'REPRO'
repro_macro = { path = "../repro_macro" }
REPRO

cat >repro_macro/src/lib.rs <<'REPRO'
extern crate proc_macro;
use proc_macro::TokenStream;

#[proc_macro_derive(repro)]
pub fn repro(_input: TokenStream) -> TokenStream {
    "".parse().unwrap()
}
REPRO

cat >>repro_macro/Cargo.toml <<'REPRO'
[lib]
proc-macro = true
REPRO

cargo build --manifest-path repro/Cargo.toml
```

---

Mentioning @petrochenkov because this bug was exposed by #51726.

### PR DESCRIPTION
**Pull request by petrochenkov**
This PR continues the work started in https://github.com/rust-lang/rust/pull/62042.
It contains a set of more or less related refactorings with the general goal of making things simpler and more orthogonal.
Along the way most of the issues uncovered in https://github.com/rust-lang/rust/pull/62086 are fixed.

The PR is better read in per-commit fashion with whitespace changes ignored.
I tried to leave some more detailed commit messages describing the motivation behind the individual changes.

Fixes https://github.com/rust-lang/rust/issues/44692
Fixes https://github.com/rust-lang/rust/issues/52363
Unblocks https://github.com/rust-lang/rust/pull/62086
r? @matthewjasper

### CODE REVIEW

**Comment by petrochenkov** [comment]
cc @eddyb @davidtwco

**Review by pnkfelix — COMMENTED** [review]
I found this note in the commit message (cec77d2ba093d108eadc95dfd2ccd0f84991cd9a) interesting:
> it was called before all the data required for resolution is available, so it could work incorrectly in some corner cases (like user-defined derives name `Copy` or `Eq`)

Can you make some regression tests illustrating these cases that used to fail, and add them either in a later commit in this PR, or in a follow-up PR?

**Comment by petrochenkov** [comment]
@pnkfelix 
>Can you make some regression tests illustrating these cases that used to fail

Added the test.

**Review by matthewjasper — APPROVED** [review]
r=me after rebase and my question answered

File: src/test/ui/hygiene/stdlib-prelude-from-opaque-early.rs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+#![feature(prelude_import)]
+
+extern crate stdlib_prelude;
+
+#[prelude_import]
+use stdlib_prelude::*;
+
+macro mac() {
+    mod m {
+        use std::mem; // OK (extern prelude)
```
[the remark this answers is not in the document]
Comment (petrochenkov) [reply src/test/ui/hygiene/stdlib-prelude-from-opaque-early.rs]: Yes, this no longer works.

Prelude names exist at the whole crate level and are supposed to look the same from every its point, macro expanded or not.
This means no one can see prelude entries introduced by opaque macros.

[the remark this answers is not in the document]
Comment (petrochenkov) [reply src/test/ui/hygiene/stdlib-prelude-from-opaque-early.rs]: I've added a test.

**Pull request merged**